File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/splashscreen/SplashScreenViewModel.java
package com.mlsdev.mlsdevstore.presentaion.splashscreen;

import android.databinding.ObservableBoolean;

import com.mlsdev.mlsdevstore.data.local.SharedPreferencesManager;
import com.mlsdev.mlsdevstore.data.model.authentication.AppAccessToken;
import com.mlsdev.mlsdevstore.data.remote.BaseObserver;
import com.mlsdev.mlsdevstore.data.remote.RemoteDataSource;
import com.mlsdev.mlsdevstore.presentaion.viewmodel.BaseViewModel;

import java.util.Calendar;

import javax.inject.Inject;

import static com.mlsdev.mlsdevstore.data.local.SharedPreferencesManager.Key;

public class SplashScreenViewModel extends BaseViewModel {
    public final ObservableBoolean appAccessTokenValid = new ObservableBoolean(false);
    private RemoteDataSource source;
    private SharedPreferencesManager preferencesManager;

    @Inject
    public SplashScreenViewModel(RemoteDataSource source, SharedPreferencesManager preferencesManager) {
        this.source = source;
        this.preferencesManager = preferencesManager;
    }

    public void checkAuthentication() {
        AppAccessToken token = preferencesManager.get(Key.APPLICATION_ACCESS_TOKEN, AppAccessToken.class);
        long currentTime = Calendar.getInstance().getTimeInMillis();
        if (token == null || (token.getExpirationDate() - currentTime) <= 0) {
            source.getAppAccessToken()
                    .subscribe(new BaseObserver<AppAccessToken>(this){
                        @Override
                        public void onSuccess(AppAccessToken accessToken) {
                            super.onSuccess(accessToken);
                            appAccessTokenValid.set(accessToken != null);
                        }
                    });
        } else {
            appAccessTokenValid.set(true);
        }
    }
}

File: app/src/main/java/com/mlsdev/mlsdevstore/presentaion/splashscreen/SplashScreenViewModel.kt
package com.mlsdev.mlsdevstore.presentaion.splashscreen

import androidx.lifecycle.MutableLiveData
import com.mlsdev.mlsdevstore.data.local.SharedPreferencesManager
import com.mlsdev.mlsdevstore.data.local.SharedPreferencesManager.Key
import com.mlsdev.mlsdevstore.data.model.authentication.AppAccessToken
import com.mlsdev.mlsdevstore.data.remote.RemoteDataSource
import com.mlsdev.mlsdevstore.presentaion.viewmodel.BaseViewModel
import java.util.*
import javax.inject.Inject

class SplashScreenViewModel @Inject
constructor(private val source: RemoteDataSource, private val preferencesManager: SharedPreferencesManager) : BaseViewModel() {
    val appAccessTokenLiveData = MutableLiveData<Boolean>()

    fun checkAuthentication() {
        val token = preferencesManager.get(Key.APPLICATION_ACCESS_TOKEN, AppAccessToken::class.java)
        val currentTime = Calendar.getInstance().timeInMillis
        if (token == null || token.expirationDate - currentTime <= 0) {
            compositeDisposable.add(source.appAccessToken.subscribe(
                    { accessToken: AppAccessToken? -> appAccessTokenLiveData.postValue(accessToken != null) },
                    { handleError(it) }))
        } else {
            appAccessTokenLiveData.postValue(true)
        }
    }
}